### PR TITLE
Add #[redb(crate = "...")] to name the redb crate in derived code

### PR DESCRIPTION
**Pull request by cberner**
Prep PR for the redb-derive audit fixes — first of the stack (this → #1311 → #1313).

The code generated by `#[derive(Value)]` and `#[derive(Key)]` referenced the redb crate by the literal path `redb`, so the derives only worked when the dependency was named `redb` at the derive site. The generated code now references the crate as `::redb` by default, and a `#[redb(crate = "...")]` attribute on the deriving struct (same shape as `#[serde(crate = "...")]`) names the crate explicitly for the remaining cases: a renamed dependency (`my_redb = { package = "redb" }`), or deriving against one of several versions of the redb package from the same crate.

Automatic rename resolution through Cargo metadata (proc-macro-crate) was in an earlier version of this PR and was dropped as not worth the complexity: it took manifest parsing plus priority rules for manifests that name several versions of the redb package, all to save one attribute at the derive site.

Tests:
- `crate_attr_tests.rs` derives for redb 2.6 and redb 4 side by side (via the attribute) and round-trips through both database versions. (The 2.6 struct is a unit struct: composed type names of field-bearing structs need `TypeName::name()`, which 2.6 keeps `pub(crate)`.)
- The `redb-derive-rename-test` workspace member (publish = false), whose only dependency on redb is under the renamed key `renamed_redb`, pins that the attribute suffices with no dependency named `redb`, and that no stray `redb` path remains in the generated code. At this point in the stack it still needs `use renamed_redb::Value as _;` for the generated code's unqualified trait-method calls — #1311 removes that requirement.

No version changes anywhere in the stack; versioning happens at release.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01EuLGiEERgJR51vhJNtYwRA

### CODE REVIEW

File: crates/redb-derive/src/lib.rs [hunk 1]
```diff
@@ -10,6 +10,21 @@ use proc_macro::TokenStream;
 use quote::quote;
 use syn::{Data, DeriveInput, Fields, GenericParam, Ident, parse_macro_input};
 
+// The path of the redb crate as seen from the deriving crate, resolved from Cargo metadata so
+// that a renamed dependency (`my_redb = { package = "redb" }`) works. `FoundCrate::Itself`
+// refers to redb's own examples, integration tests, and doctests, which link redb under its own
+// name (the library itself cannot use these derives, as that would be circular). The fallback
+// covers build systems without Cargo metadata, where the dependency must be named `redb`.
+fn redb_crate_path() -> proc_macro2::TokenStream {
+    match proc_macro_crate::crate_name("redb") {
```
[the remark this answers is not in the document]
Comment (cberner) [reply crates/redb-derive/src/lib.rs]: Confirmed and fixed. Reproduced by adding `redb2_6 = { package = "redb", version = "=2.6.0" }` after `redb` in redb-derive&#39;s own dev-dependencies: proc-macro-crate collects dependencies into a map keyed by package name, so the last declaration wins and the whole derive test suite failed to compile against `::redb2_6`, exactly as described.

The fix (now in this PR): a dependency actually named `redb` takes priority over proc-macro-crate&#39;s answer — the renamed lookup only applies when no dependency named `redb` exists, which also matches how the generated code resolved before renames were supported. The poisoned `redb2_6` dev-dependency stays in the manifest, deliberately declared last, as a regression test: it captures the entire derive test suite if the preference ever breaks.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (cberner) [reply crates/redb-derive/src/lib.rs]: The changelog for the whole stack is deliberately carried by the final PR (#1313), per maintainer preference: its `redb-derive 0.2.0` section includes a bullet for renamed-dependency support, so the release notes do cover this change. This PR and #1311 intentionally leave CHANGELOG.md untouched.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (cberner) [reply crates/redb-derive/src/lib.rs]: Is that really the right fix? What if someone wants to derive for both redb and redb2_6?

[the remark this answers is not in the document]
Comment (cberner) [reply crates/redb-derive/src/lib.rs]: Deriving for both is now supported, but it can&#39;t come from resolution: any scheme picks one crate per expansion site, since a derive emits impls of exactly one crate&#39;s traits. What it needs is per-struct control, so I&#39;ve added the standard escape hatch (same shape as `#[serde(crate = "...")]`):

```rust
#[derive(Value, Key)]
#[redb(crate = "redb2_6")]
struct OldFormat;

#[derive(Value, Key)]
#[redb(crate = "::redb")]
struct NewFormat { ... }
```

The attribute overrides all resolution; the bare-name preference is only the default for unannotated derives (where it keeps the migration pattern from this repo&#39;s own manifest working). The new `crate_attr_tests.rs` derives for redb 2.6 and redb 4 side by side in redb-derive&#39;s own test suite and round-trips through both database versions.

One caveat surfaced by that test: structs *with fields* can&#39;t target 2.6 no matter how the crate path is chosen — the generated composed `type_name()` reads field names via `TypeName::name()`, which is `pub(crate)` in 2.6 (public in 4.x). So the 2.6 side of the test is a unit struct; field-bearing derives work for any redb version that exposes `TypeName::name()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/redb-derive/src/lib.rs [hunk 2]
```diff
@@ -10,6 +10,74 @@ use proc_macro::TokenStream;
...
+    }
+}
+
+// proc-macro-crate resolves a package to the last dependency in the manifest whose package name
+// matches, so a dependency named `redb` would otherwise lose to a renamed dependency on another
+// version of the redb package declared after it, such as an old version kept alongside for
+// migration. A dependency actually named `redb` therefore takes priority, which also matches how
+// the generated code resolved before renames were supported.
+fn manifest_names_redb_dependency() -> bool {
+    const DEP_TABLES: [&str; 3] = ["dependencies", "dev-dependencies", "build-dependencies"];
```
[the remark this answers is not in the document]
Comment (cberner) [reply crates/redb-derive/src/lib.rs]: There&#39;s no reliable way to scope the scan to the current target: a proc macro gets no signal about which target is compiling. `CARGO_TARGET_TMPDIR` marks only integration tests and benches — unit tests, examples, and doctests all link dev-dependencies without it — so restricting the override to `[dependencies]` would break the far more common test-only-redb layout (including redb-derive&#39;s own suite, where the duplicate-package regression this scan fixes was reproduced). Plain `proc_macro_crate` fails this configuration identically, for the same reason: it scans dev-dependencies too, and the last package-name match wins.

The configuration is now expressible explicitly instead: `#[redb(crate = "renamed_redb")]` on the deriving struct bypasses the scan entirely.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/redb-derive/src/lib.rs [hunk 3]
```diff
@@ -66,14 +102,15 @@ fn generate_value_impl(input: &DeriveInput) -> syn::Result<proc_macro2::TokenStr
     let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
 
     let self_type = generate_self_type(name, generics)?;
+    let redb = redb_crate_path_for(input)?;
 
-    let type_name_impl = generate_type_name(name, &data_struct.fields);
+    let type_name_impl = generate_type_name(name, &data_struct.fields, &redb);
     let as_bytes_impl = generate_as_bytes(&data_struct.fields);
     let from_bytes_impl = generate_from_bytes(name, &data_struct.fields);
     let fixed_width_impl = generate_fixed_width(&data_struct.fields);
```
[the remark this answers is not in the document]
Comment (cberner) [reply crates/redb-derive/src/lib.rs:110]: Correct about this PR in isolation, and it's exactly what the next PR in the stack does: #1311 rewrites the generated code to spell every path fully qualified, threading the crate path selected here through all the field helpers — `<Field as #redb::Value>::fixed_width/as_bytes/from_bytes` and the field type-name calls included.

This PR deliberately leaves the field helpers' pre-existing unqualified calls untouched (one concern per PR), which is why its tests still bring exactly one `Value` trait into scope per module (`use redb::Value as _;` / `use redb2_6::Value as _;`) — avoiding the ambiguity you describe — and #1311 drops those imports in the same commit that qualifies the calls. Deriving field-bearing structs against two versions in one module works from #1311 onward.

---
_Generated by [Claude Code](https://claude.ai/code)_
